metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,395 @@
+#include "dbg_sshash.hpp"
+
+#include <type_traits>
+
+#include "common/seq_tools/reverse_complement.hpp"
+#include "common/threads/threading.hpp"
+#include "common/logger.hpp"
+#include "kmer/kmer_extractor.hpp"
+
+
+namespace mtg {
+namespace graph {
+
+#if _PROTEIN_GRAPH
+static constexpr uint16_t bits_per_char = sshash::aa_uint_kmer_t<uint64_t>::bits_per_char;
+#else
+static constexpr uint16_t bits_per_char = sshash::dna_uint_kmer_t<uint64_t>::bits_per_char;
+#endif
+
+template <typename T>
+struct template_parameter;
+
+template <template <typename ...> class C, typename T>
+struct template_parameter<C<T>> { using type = T; };
+
+template <typename T>
+using get_kmer_t = typename template_parameter<typename std::remove_const<typename std::remove_reference<T>::type>::type>::type;
+
+const std::string DBGSSHash::alphabet_ = kmer::KmerExtractor2Bit().alphabet;
+
+constexpr DeBruijnGraph::node_index sshash_to_graph_index(uint64_t idx) {
+    return idx + 1;
+}
+constexpr uint64_t graph_index_to_sshash(DeBruijnGraph::node_index idx) {
+    return idx - 1;
+}
+
+size_t DBGSSHash::dict_size() const {
+    return std::visit([](const auto &d) { return d.size(); }, dict_);
+}
+
+DBGSSHash::DBGSSHash(size_t k, Mode mode) : k_(k), num_nodes_(0), mode_(mode) {
+    size_t odd_k = (k_ | 1);
+
+    if (odd_k * bits_per_char <= 64) {
+        dict_ = sshash::dictionary<kmer_t<KmerInt64>>();
+    } else if (odd_k * bits_per_char <= 128) {
+        dict_ = sshash::dictionary<kmer_t<KmerInt128>>();
+    } else if (odd_k * bits_per_char <= 256) {
+        dict_ = sshash::dictionary<kmer_t<KmerInt256>>();
+    } else {
+        common::logger->error("k too big: {} > {}", odd_k, 256 / bits_per_char);
+        throw std::length_error("k fail");
+    }
+}
+
+DBGSSHash::DBGSSHash(const std::string &input_filename, size_t k, Mode mode, size_t num_chars)
+      : DBGSSHash(k, mode) {
+    sshash::build_configuration build_config;
+    build_config.k = k;
+
+    // use a value of m recommended here:
+    // https://twitter.com/giulio_pibiri/status/1803417277213114501
+    // otherwise, use the next odd value greater than or equal to floor(k / 2)
+    build_config.m = std::min(
+        num_chars > 0
+            ? uint64_t(ceil(log(static_cast<double>(num_chars)) / log(static_cast<double>(alphabet_.size()))) + 1)
+            : uint64_t(k / 2) | 1,
+        sshash::constants::max_m
+    );
```

Comment:
Should we do `build_config.m |= 1` after initial assignment? otherwise the first case may accidentally give even `m`, and I'm not sure if sshash is good with them.